### ISSUE TITLE
Building simh with Xcode reveals many complaints ..

### ISSUE THREAD

**Issue by gavineadie** [simh/simh#846]
- #### Context
This may be an issue that is more trouble than it is worth .. in which case just mark it closed.  I can subscribe to the motto "if it ain't broke, don't fix it" and withdraw back to lurking without rancor.

Because I wanted to find out more about the operation of simh, I built the PDP-11 simulator with the macOS Xcode IDE (Version 11.4.1) in order to set breakpoint and examine variables.  The clang compiler has some powerful tools to catch potential code problems and to analyze program flow .. Xcode uses these and, as a result, while there are no "errors" and simh runs correctly, there are ~170 "warnings" and ~110 "logic errors" reported.

I have not taken time to chase down any of the issues raised, but I will if deemed worth it and to provide a sense of how much trouble it would be to address these.

I recognize that the simh code is given tender loving care, I am in no way being critical, and it runs beautifully to do what I (and everyone else) want.  Moreover, I am not letting the clever makefile perform all its tricks to test conditions and set definitions, so this compilation is a worst case scenario.

Would there any merit to my sharing all (or a carefully selected, genuine few) of the warnings and logic "errors" that clang is coughing up? 

- #### the output of "sim> SHOW VERSION" while running the simulator which is having the issue
    sim> show version
        PDP-11 simulator V4.0-0 Current
            Simulator Framework Capabilities:
        32b data
        32b addresses
        no Ethernet
        Idle/Throttling support is available
        Virtual Hard Disk (VHD) support
        RAW disk and CD/DVD ROM support
        FrontPanel API Version 12
    Host Platform:
        Compiler: GCC 4.2.1 Compatible Apple LLVM 11.0.3 (clang-1103.0.32.59)
        Simulator Compiled as C arch: x64 (Debug Build) on May  2 2020 at 00:52:01
        Build Tool: undefined (probably cmake)
        Memory Access: Little Endian
        Memory Pointer Size: 64 bits
        Large File (>2GB) support
        SDL Video support: No Video Support
        No RegEx support for EXPECT commands
        OS clock resolution: 1ms
        Time taken by msleep(1): 1ms
        OS: Darwin Ramsay-Mobile.attlocal.net 19.4.0 Darwin Kernel Version 19.4.0: Wed Mar  4 22:28:40 PST 2020; root:xnu-6153.101.6~15/RELEASE_X86_64 x86_64
        git commit id: $Format:
        git commit time: $Format:%aI$

Notice that not using the makefile is resulting in some information lacking (like git info).  I was using the master branch from a few days ago.
- #### how you built the simulator or that you're using prebuilt binaries
Xcode .. as described above.

- #### the simulator configuration file (or commands) which were used when the problem occurred.
I'm not running simh, just compiling it.

- #### the expected behavior and the actual behavior
I'm not running simh, just compiling it.

- #### you may also need to provide specific pointers to data files that may be necessary to demonstrate the problem
A file is attached that summarizes the compiler's complaints.  There's a lot, but I provide it for examples of what is being caught.

[simh+xcode.txt](https://github.com/simh/simh/files/4567142/simh%2Bxcode.txt)

**Comment by markpizz** [comment]
I've got a MacOS test/build environment with XCode installed.  How do I reproduce what you've done here?  If you can attach anyfiles you may have setup (zipped) to this ticket, that would be easiest for me.

**Comment by gavineadie** [comment]
Apologies .. I didn't see this till now.  I'll get you something over the weekend ..

**Comment by gavineadie** [comment]
I'll break some weekend discoveries into a few "Issues" .. I've got some documentation corrections I'll do in a new issue.

Staying on this issue's topic, I'll give one easy example of clang's persnickety behavior.  In the file `pop11_vh.c` (around line 838), is the function:

```
static int32 tx_fifo_put ( TMLX    *lp,
            int32   data    )
{
    int32   status = 0;

    if (tx_fifo_free_count (lp) == 0)
        return -1;
    lp->txfifo[(lp->txfifo_idx + lp->txfifo_cnt) % TX_FIFO_SIZE] = data;
    ++lp->txfifo_cnt;
    return 0;
}
```

.. and clang notices that the variable `status` is never used.  There are 12 such complaints in an Xcode compilation of pdp11.

Clang also catches "*Variable 'rtc' may be uninitialized when used here, Initialize the variable 'rtc' to silence this warning*" and "*Code will never be executed, Silence by adding parentheses to mark code as explicitly dead*"

Neither of the above "Unused Entity" or "Semantic" issues is likely to be critical, but there are also "Value Conversion" issues, like "*Implicit conversion loses integer precision: 'unsigned long' to 'int'*" and "*Implicit conversion loses integer precision: 'ssize_t' (aka 'long') to 'int'*" that might actually cause a computational error (though going from 32-bit to 64-bit, is much safer than in the other direction).

What I will do is send you (attached here, if it fits) my Xcode project file. You'll have to adjust where it thinks the sources are to be found (two groups: "simh" and "pdp11"), otherwise let me know if you have trouble.

If you want I can create a PR for the "Unused Entity" and "Semantic" issues .. that's easy!  The "Value Conversion" issues need much more careful examination before being changed for all the platforms simh is built on.

[pdp11.xcodeproj.zip](https://github.com/simh/simh/files/4783256/pdp11.xcodeproj.zip)

**Comment by gavineadie** [comment]
If you set clang's static analyzer loose on the code more worrying things appear.  For example, in `pdp11_tq.c` around line 1683:
```
t_bool tq_hbe (UNIT *uptr, uint32 ba)
{
uint16 pkt, tpkt;

if ((tq_cflgs & CF_THS) == 0)                           /* logging? */
    return OK;
if (!tq_deqf (&pkt))                                    /* get log pkt */
    return ERR;
tpkt = uptr->cpkt;                                      /* rw pkt */
tq_pkt[pkt].d[ELP_REFL] = tq_pkt[tpkt].d[CMD_REFL];     /* copy cmd ref */
tq_pkt[pkt].d[ELP_REFH] = tq_pkt[tpkt].d[CMD_REFH];     /* copy cmd ref */
tq_pkt[pkt].d[ELP_UN] = tq_pkt[tpkt].d[CMD_UN];         /* copy unit */
. . . 
```
.. the static analyzer yelps about `pkt` ..  "Array subscript is undefined"

I do have to note that I am not examining these clang complaints in detail, some are very complicated reaching over dozens of lines of code, and I'm not familiar enough with simh to determine if the analyzer is just getting confused.  I'm just raising awareness .. as I said way back when I started this, I can subscribe to the motto "if it ain't broke, don't fix it" ..

But if going down this rabbit hole is deemed worth it, I'll contribute to fixing things.

**Comment by markpizz** [comment]
Well, usually we welcome more compiler generate feedback and certainly strive to fix anything that compilers explicitly complain about when invoked with default compile arguments.  Different compilers find different things, so once again more is usually better.
On the other hand, if a tool (compiler or other analyzer tool) produces 99% noise and no useful info, we'll walk away from it pretty fast.

Not seeing the precise wording of the complaint about pkt in tq_hbe(), I'm not convinced this is good advice or merely noise.  Yes, there is a path in tq_deqf() that doesn't set the value of pkt, but when that happens tq_deqf() returns a status which causes tq_hbe() to return ERR and thus never use the uninitialized value of pkt.

The codebase is periodically passed through Coverity's static analysis, and when reasonable and possible we change the code to address what has been pointed out.  Some cases are real false positives and Coverity is good at letting things get classified as such.  

I'll get to looking closer at this eventually.  Thanks.

**Comment by larsbrinkhoff** [comment]
Speaking of CI builds, I no longer see Travis CI reporting back to GitHub.  But I do see the builds here: https://travis-ci.org/github/simh/simh

**Comment by gavineadie** [comment]
@markpizz .. "I'll get to looking closer at this eventually."  

Fine, I'm happy to help when you do; meanwhile, silence!

Just for completeness, and not denying your good point about noise, the analyzer's full output (which I didn't look at, and should have, when selecting an example, mea culpa) explicitly states the 17 steps from `tq_svc` via `tq_deqf` to the point of concern in `tq_hbe`, requiring at least 7 conditions to be true.  It blindly searches for a weak path not knowing that at least of one those conditions has to be false by, I assume, the state of the emulated device.

**Comment by eschaton** [comment]
Hey there @gavineadie, it's been a while! Glad to see you around. :) I also see tons of warnings building SIMH in Xcode with the config I mention in #697, and virtually all of them appear substantial. I wish I could share my projects to make it easier for folks to replicate, but I can describe a little bit more here:

- I created an Xcode Projects directory next to the Visual Studio Projects one, and have an Xcode workspace in it for SIMH.
- I use a shared xcconfig that establishes all of the common preprocessor macros to use for all simulators. I do not include this in any simulator projects directly though.
- I have an A32I32 xcconfig and an A64I64 xcconfig that include the shared xcconfig and specify the preprocessor macros to use for A32/I32 and A64/I64 variants of SIMH.
- I made a project in the workspace to build a pair of "libsimh" static libraries, libsimh_a32i32.a and libsimh_a64i64.a, whose targets are each based on the appropriate xcconfig. I put all of the common files into this.
- I made a project and command line tool target in it for each simulator I care about, containing only its sources
  - Each target is based on either the A32I32 or A64I64 xcconfig, as appropriate. (E.g. PDP11 is A32I32.)
  - Each target links the appropriate libsimh_*.a static library

This has worked quite well and made it very easy to do things like try out changes, run the static analyzer, and so on.
